src/server/main_service.cc

```diff
@@ -879,35 +918,41 @@ void Service::EvalInternal(const EvalArgs& eval_args, Interpreter* interpreter,
 void Service::Discard(CmdArgList args, ConnectionContext* cntx) {
   RedisReplyBuilder* rb = (*cntx).operator->();
 
-  if (cntx->conn_state.exec_state == ConnectionState::EXEC_INACTIVE) {
+  if (!cntx->conn_state.exec_info.IsActive()) {
     return rb->SendError("DISCARD without MULTI");
   }
 
-  cntx->conn_state.exec_state = ConnectionState::EXEC_INACTIVE;
-  cntx->conn_state.exec_body.clear();
-
+  cntx->conn_state.exec_info.Clear();
   rb->SendOk();
 }
 
 void Service::Exec(CmdArgList args, ConnectionContext* cntx) {
   RedisReplyBuilder* rb = (*cntx).operator->();
 
-  if (cntx->conn_state.exec_state == ConnectionState::EXEC_INACTIVE) {
+  if (!cntx->conn_state.exec_info.IsActive()) {
     return rb->SendError("EXEC without MULTI");
   }
 
-  if (cntx->conn_state.exec_state == ConnectionState::EXEC_ERROR) {
-    cntx->conn_state.exec_state = ConnectionState::EXEC_INACTIVE;
-    cntx->conn_state.exec_body.clear();
+  if (cntx->conn_state.exec_info.state == ConnectionState::ExecInfo::EXEC_ERROR) {
+    cntx->conn_state.exec_info.Clear();
     return rb->SendError("-EXECABORT Transaction discarded because of previous errors");
   }
 
-  VLOG(1) << "StartExec " << cntx->conn_state.exec_body.size();
-  rb->StartArray(cntx->conn_state.exec_body.size());
-  if (!cntx->conn_state.exec_body.empty()) {
+  auto& exec_info = cntx->conn_state.exec_info;
+
+  // TODO: check watched keys for expiry
+  if (exec_info.watched_dirty.load()) {
+    exec_info.Clear();
+    return rb->SendNull();
+  }
```

Comment:
See my TODO.

If watched is dirty, then a missing key just means we're looking at the wrong shard? Or should we filter them (again filter issue)